reactive/kotlinx-coroutines-rx2/test/SchedulerTest.kt

```diff
@@ -31,4 +38,252 @@ class SchedulerTest : TestBase() {
         }
         finish(4)
     }
+
+    @Test
+    fun testAsSchedulerWithNoDelay(): Unit = runTest {
+        expect(1)
+        val mainThread = Thread.currentThread()
+        val scheduler = (currentDispatcher() as CoroutineDispatcher).asScheduler()
+        suspendCancellableCoroutine<Unit> {
+            scheduler.scheduleDirect {
+                val t1 = Thread.currentThread()
+                assertSame(t1, mainThread)
+                expect(2)
+                it.resume(Unit)
+            }
+        }
+        yield()
+        finish(3)
+    }
+
+    @Test
+    fun testAsSchedulerWithDelay(): Unit = runTest {
+        expect(1)
+        val mainThread = Thread.currentThread()
+        val scheduler = (currentDispatcher() as CoroutineDispatcher).asScheduler()
+        val delayMillis = 300L
+        suspendCancellableCoroutine<Unit> {
+            scheduler.scheduleDirect({
+                val t1 = Thread.currentThread()
+                assertSame(t1, mainThread)
+                expect(2)
+                it.resume(Unit)
+            }, delayMillis, TimeUnit.MILLISECONDS)
+        }
+        finish(3)
+    }
+
+    @Test
+    fun testAsSchedulerWithZeroDelay(): Unit = runTest {
+        expect(1)
+        val mainThread = Thread.currentThread()
+        val scheduler = (currentDispatcher() as CoroutineDispatcher).asScheduler()
+        suspendCancellableCoroutine<Unit> {
+            scheduler.scheduleDirect({
+                val t1 = Thread.currentThread()
+                assertSame(t1, mainThread)
+                expect(2)
+                it.resume(Unit)
+            }, 0, TimeUnit.MILLISECONDS)
+        }
+
+        scheduler.shutdown()
+        finish(3)
+    }
+
+    @Test
+    fun testAsSchedulerWithNegativeDelay(): Unit = runTest {
+        expect(1)
+        val mainThread = Thread.currentThread()
+        val scheduler = (currentDispatcher() as CoroutineDispatcher).asScheduler()
+        suspendCancellableCoroutine<Unit> {
+            scheduler.scheduleDirect({
+                val t1 = Thread.currentThread()
+                assertSame(t1, mainThread)
+                expect(2)
+                it.resume(Unit)
+            }, -1, TimeUnit.MILLISECONDS)
+        }
+        yield()
+        finish(3)
+    }
+
+    @Test
+    fun testDisposeDuringDelay(): Unit = runTest {
+        expect(1)
+        val scheduler = (currentDispatcher() as CoroutineDispatcher).asScheduler()
+        val delayMillis = 300L
+        val disposable = scheduler.scheduleDirect({
+            expectUnreached()
+        }, delayMillis, TimeUnit.MILLISECONDS)
+        delay(100)
```

Comment:
These tests should be using a `DelayController` to advance time instead of actually sleeping.